File: app/src/main/java/com/bogdan/kolomiiets/birthdayreminder/MainActivity.java
package com.bogdan.kolomiiets.birthdayreminder;

import android.content.ContentValues;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.os.Bundle;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import android.view.ContextMenu;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.SimpleAdapter;
import android.widget.TextView;
import android.widget.Toast;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.MobileAds;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Locale;
/*
public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private SQLiteDBHelper dbHelper;
    private SQLiteDatabase db;
    private FloatingActionButton fab;
    private RelativeLayout relativeLayout;
    private ListView lv_dynamic;
    private int mItemId = 0;
    private String mItemName = "";
    private String mItemPhone = "";
    private AdView mAdView;
    public static InterstitialAd mInterstitialAd;
    private SharedPreferences preferences;
    public static int countForShowAd;
    public static final int clickCountForShowAd = 5;
    private ListAdapter adapter;
    private String selection = SQLiteDBHelper.KEY_NAME + " like ?";
    private String[] selectionArgs = {"%%"};
    private String searchQuery;
    private MenuItem item;
    private SearchView searchView;
    private TextView noData;
    private ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(this);
        fab.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                fab.requestFocus();
                return false;
            }
        });


        //Setup default sharedPreferences
        initSharedPreferences();

        //init searchQuery
        if (savedInstanceState != null)
        {
            searchQuery = savedInstanceState.getString("searchQuery");
        }

        relativeLayout = (RelativeLayout) findViewById(R.id.relativeLayout);

        //create TextView noData for empty ListView
        RelativeLayout.LayoutParams progressParams = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
        progressParams.addRule(RelativeLayout.CENTER_IN_PARENT);
        progressBar = new ProgressBar(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this);
        progressBar.setVisibility(View.INVISIBLE);
        progressBar.setLayoutParams(progressParams);

        RelativeLayout.LayoutParams noDataParams = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.MATCH_PARENT);
        noData = new TextView(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this);
        noData.setText(R.string.nothing_to_show);
        noData.setGravity(Gravity.CENTER);
        noData.setTextSize(getResources().getDimension(R.dimen.size12));
        noDataParams.addRule(RelativeLayout.CENTER_HORIZONTAL);
        noData.setLayoutParams(noDataParams);

        relativeLayout.addView(noData);
        relativeLayout.addView(progressBar);

        lv_dynamic = (ListView) findViewById(R.id.lv_dynamic);

        mAdView = findViewById(R.id.adView);
        MobileAds.initialize(this, getResources().getString(R.string.app_id));

        //BannerAd
        AdRequest adRequest = new AdRequest.Builder().build();
        //adRequest.isTestDevice(this);
        mAdView.loadAd(adRequest);

        //InterstitialAd
        mInterstitialAd = new InterstitialAd(this);
        mInterstitialAd.setAdUnitId("ca-app-pub-8123334718162410/4114864477");
        mInterstitialAd.loadAd(new AdRequest.Builder().build());
        mInterstitialAd.setAdListener(new AdListener(){
            @Override
            public void onAdOpened() {
                countForShowAd = 0;
            }

            @Override
            public void onAdClosed() {
                mInterstitialAd.loadAd(new AdRequest.Builder().build());
            }
        });

    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putStringArray("selectionArgs", selectionArgs);
        outState.putString("searchQuery", searchView.getQuery().toString());
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        selectionArgs = savedInstanceState.getStringArray("selectionArgs");
    }

    @Override
    protected void onResume() {
        super.onResume();
        QueryToDB();
    }

    //Setup default sharedPreferences
    private void initSharedPreferences() {
        preferences = getSharedPreferences(SettingsActivity.PREFERENCES_NAME, MODE_PRIVATE);
        if (!preferences.contains(SettingsActivity.SWITCH)) {
            SharedPreferences.Editor editor = preferences.edit();
            editor.putBoolean(SettingsActivity.SWITCH, true);
            editor.putInt(SettingsActivity.CHECKED_RADIO, 0);
            editor.putInt(SettingsActivity.REMINDER_HOUR, 9);
            editor.putInt(SettingsActivity.REMINDER_MINUTE, 0);
            editor.commit();

            //Set alarm
            SetAlarm setAlarm = new SetAlarm();
            setAlarm.onAlarm(this,true, 9, 0);
        }
    }

    public void onAboutMenuItem(MenuItem item) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this);
        builder.setTitle(R.string.about_the_program);
        builder.setIcon(R.mipmap.ic_launcher);
        View view = getLayoutInflater().inflate(R.layout.about_layout, null);
        builder.setView(view);
        TextView feedback = view.findViewById(R.id.email);
        feedback.setOnClickListener(this);

        builder.setPositiveButton("FEEDBACK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                Intent intent = new Intent(Intent.ACTION_SENDTO);
                intent.setData(Uri.parse("mailto: bogdan.kolomiiets@gmail.com"));
                startActivity(Intent.createChooser(intent, "Send feedback"));
            }
        });
        builder.setNegativeButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });
        builder.create();
        builder.show();
    }

    public void onSettingsMenuItem(MenuItem item) {
        Intent intent = new Intent(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, SettingsActivity.class);
        startActivity(intent);
    }

    public void onExportData(MenuItem item){
        ExportDataFromDatabase data_export = new ExportDataFromDatabase(this);
        data_export.execute();
    }

    public void onImportData(MenuItem item){
        Intent selectFileIntent = null;
        /*if (Build.VERSION.SDK_INT >= 19){
            selectFileIntent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            selectFileIntent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, false);
            selectFileIntent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
        } else {*//*
            selectFileIntent = new Intent(Intent.ACTION_GET_CONTENT);
        //}
        //selectFileIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        //selectFileIntent.setType("");
        selectFileIntent.setType("application/octet-stream");
        startActivityForResult(selectFileIntent, RequestCodes.SELECT_FILE);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_item, menu);
        item = menu.findItem(R.id.search);
        searchView = (SearchView) item.getActionView();
        if (searchQuery != null && !searchQuery.isEmpty())
        {
            searchView.setQuery(searchQuery, false);
            searchView.setIconified(false);
        }
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                selectionArgs = new String[] {"%" + newText.toLowerCase().trim() + "%"};
                QueryToDB();
                return true;
            }
        });
    return true;
    }

    public void onContextChange(MenuItem item){
        Intent intent = new Intent(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, EditCelebrationActivity.class);
        intent.putExtra("id", mItemId);
        startActivity(intent);
        finish();
    }

    public void onContextDelete(MenuItem item){
        AlertDialog.Builder builder = new AlertDialog.Builder(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this);
        builder.setTitle(R.string.are_you_sure);
        builder.setCancelable(true);
        builder.setMessage(getResources().getString(R.string.really_delete) + " " + mItemName + "\n" + mItemPhone);
        builder.setNegativeButton(R.string.negative_answer, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

            }
        });
        builder.setPositiveButton(R.string.positive_answer, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                DeleteRowFromDatabase delete = new DeleteRowFromDatabase(mItemId, com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this);
                if (delete.deleteRow()){
                    Toast.makeText(getApplicationContext(), getResources().getText(R.string.deleted), Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getApplicationContext(), getResources().getText(R.string.not_deleted), Toast.LENGTH_SHORT).show();
                }
                QueryToDB();
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void QueryToDB() {
        progressBar.setVisibility(View.VISIBLE);
        new Thread() {
            @Override
            public void run() {
                try {
                    dbHelper = new SQLiteDBHelper(getApplicationContext());
                    db = dbHelper.getReadableDatabase();
                    Cursor cursor = db.query(SQLiteDBHelper.TABLE_NAME, SQLiteDBHelper.COLUMNS_NAMES, selection, selectionArgs, null, null, null);
                    final ArrayList<DataForListView> list = new ArrayList<>();

                    if (cursor.getCount() == 0) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                noData.setVisibility(View.VISIBLE);
                            }
                        });
                    } else {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                noData.setVisibility(View.INVISIBLE);

                            }
                        });
                        cursor.moveToFirst();
                        do {
                            int id = cursor.getInt(cursor.getColumnIndex(SQLiteDBHelper.KEY_ID));
                            String name = cursor.getString(cursor.getColumnIndex(SQLiteDBHelper.KEY_NAME));
                            String phone = cursor.getString(cursor.getColumnIndex(SQLiteDBHelper.KEY_PHONE));

                            String type = "";
                            switch (cursor.getInt(cursor.getColumnIndex(SQLiteDBHelper.KEY_CELEBRATION_TYPE))) {
                                case 1:
                                    type = getResources().getString(R.string.birthday) + ":";
                                    break;
                                case 2:
                                    type = getResources().getString(R.string.anniversary) + ":";
                                    break;
                                case 3:
                                    type = getResources().getString(R.string.holiday) + ":";
                                    break;
                            }

                            String date;
                            int year = cursor.getInt(cursor.getColumnIndex(SQLiteDBHelper.KEY_YEAR));
                            int month = cursor.getInt(cursor.getColumnIndex(SQLiteDBHelper.KEY_MONTH));
                            int day = cursor.getInt(cursor.getColumnIndex(SQLiteDBHelper.KEY_DAY));

                            String age = "";
                            //Check if it's holiday, then skip year
                            if (year > 1) {
                                if (Locale.getDefault().getLanguage().equals("ru") || Locale.getDefault().getLanguage().equals("uk")) {
                                    date = day + "/" + month + "/" + year;
                                } else date = year + "-" + month + "-" + day;
                                //Calculate age
                                age = String.valueOf(new HowOld().getHowOld(year, month, day));
                            } else {
                                if (Locale.getDefault().getLanguage().equals("ru") || Locale.getDefault().getLanguage().equals("uk")) {
                                    date = day + "/" + month;
                                } else date = month + "-" + day;
                            }



                            list.add(new DataForListView(String.valueOf(id), name, phone, type, age, date));
                        } while (cursor.moveToNext());
                    }

                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            adapter = new SimpleAdapter(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, list, R.layout.not_simple_item, new String[]
                                    {SQLiteDBHelper.KEY_ID, SQLiteDBHelper.KEY_NAME, SQLiteDBHelper.KEY_PHONE, SQLiteDBHelper.KEY_CELEBRATION_TYPE, "date", "age"},
                                    new int[]{R.id.tv_id, R.id.name_container, R.id.phone_container, R.id.type_container, R.id.date_container, R.id.age_container});
                            lv_dynamic.setAdapter(adapter);
                            lv_dynamic.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                                @Override
                                public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                                    TextView mTv_id = (TextView) view.findViewById(R.id.tv_id);
                                    TextView mName_container = (TextView) view.findViewById(R.id.name_container);
                                    TextView mPhone_container = (TextView) view.findViewById(R.id.phone_container);
                                    mItemId = Integer.parseInt(mTv_id.getText().toString());
                                    mItemName = mName_container.getText().toString();
                                    mItemPhone = mPhone_container.getText().toString();
                                    return false;
                                }
                            });
                            lv_dynamic.setOnCreateContextMenuListener(new View.OnCreateContextMenuListener() {
                                @Override
                                public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
                                    MenuInflater inflater = new MenuInflater(v.getContext());
                                    inflater.inflate(R.menu.popup_menu, menu);
                                }
                            });
                        }
                    });
                    cursor.close();
                } catch (final Exception e) {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
                } finally {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            progressBar.setVisibility(View.INVISIBLE);
                        }
                    });
                    db.close();
                }
            }
        }.start();
    }

    @Override
    public void onClick(View v) {
        Intent intent;
        switch (v.getId()){
            case R.id.fab:
                countForShowAd++;
                if (countForShowAd > clickCountForShowAd){
                    if (mInterstitialAd.isLoaded()){
                        mInterstitialAd.show();
                    }
                }
                intent = new Intent(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, NewCelebrationActivity.class);
                startActivity(intent);
                break;
            case R.id.email:
                intent = new Intent(Intent.ACTION_SENDTO);
                intent.setData(Uri.parse("mailto: bogdan.kolomiiets@gmail.com"));
                startActivity(Intent.createChooser(intent, "Send feedback"));
                break;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RequestCodes.SELECT_FILE && resultCode == RESULT_OK){
            Uri uri = data.getData();
            importDataToDatabase(uri);
        }
    }

    private void importDataToDatabase(final Uri uri) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                JSONArray jsonArray;
                JSONObject jsonObject;
                ContentValues contentValues;
                SQLiteDBHelper helper;
                SQLiteDatabase database;
                BufferedReader bufferedReader = null;
                boolean import_result = false;

                helper = new SQLiteDBHelper(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this);
                database = helper.getWritableDatabase();
                contentValues = new ContentValues();
                try {
                    bufferedReader = new BufferedReader(new InputStreamReader(new BufferedInputStream(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this.getContentResolver().openInputStream(uri))));
                    StringBuilder stringBuilder = new StringBuilder();
                    String line = "";
                    while ((line = bufferedReader.readLine()) != null) {
                        stringBuilder.append(line).append("\n");
                    }
                    try {
                        jsonArray = new JSONArray(stringBuilder.toString());
                        database.beginTransaction();
                        for (int i = 0; i < jsonArray.length(); i++) {
                            jsonObject = jsonArray.getJSONObject(i);
                            contentValues.put(SQLiteDBHelper.KEY_NAME, jsonObject.getString(SQLiteDBHelper.KEY_NAME));
                            contentValues.put(SQLiteDBHelper.KEY_PHONE, jsonObject.getString(SQLiteDBHelper.KEY_PHONE));
                            contentValues.put(SQLiteDBHelper.KEY_CELEBRATION_TYPE, jsonObject.getInt(SQLiteDBHelper.KEY_CELEBRATION_TYPE));
                            contentValues.put(SQLiteDBHelper.KEY_YEAR, jsonObject.getInt(SQLiteDBHelper.KEY_YEAR));
                            contentValues.put(SQLiteDBHelper.KEY_MONTH, jsonObject.getInt(SQLiteDBHelper.KEY_MONTH));
                            contentValues.put(SQLiteDBHelper.KEY_DAY, jsonObject.getInt(SQLiteDBHelper.KEY_DAY));
                            database.insert(SQLiteDBHelper.TABLE_NAME, null, contentValues);
                        }
                        database.setTransactionSuccessful();
                        database.endTransaction();
                        import_result = true;
                    } catch (JSONException e) {
                        e.printStackTrace();
                        import_result = false;
                    }
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    database.close();
                    if (bufferedReader != null) {
                        try {
                            bufferedReader.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {

                    }
                });
                if (import_result) {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, getApplicationContext().getResources().getString(R.string.imported_successful), Toast.LENGTH_SHORT).show();
                            QueryToDB();
                        }
                    });
                } else if (!import_result) {
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(com.bogdan.kolomiiets.birthdayreminder.views.MainActivity.this, getApplicationContext().getResources().getString(R.string.not_imported), Toast.LENGTH_SHORT).show();

                        }
                    });
                }
            }
        }).start();
    }
}
*/

File: app/src/main/java/com/bogdan/kolomiiets/birthdayreminder/views/MainActivity.kt
package com.bogdan.kolomiiets.birthdayreminder.views

import android.Manifest.permission.READ_EXTERNAL_STORAGE
import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.core.app.ActivityCompat
import androidx.core.content.PermissionChecker
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.BASE_REMINDER_HOUR
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.BASE_REMINDER_MINUTE
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.CHECKED_RADIO
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.EVENT
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.IS_REMINDER_ON
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.PREFERENCES_NAME
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.RADIO_SAME_DAY
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.REMINDER_HOUR
import com.bogdan.kolomiiets.birthdayreminder.Constants.Companion.REMINDER_MINUTE
import com.bogdan.kolomiiets.birthdayreminder.R
import com.bogdan.kolomiiets.birthdayreminder.RequestCodes.Companion.INTENT_ACTION_CREATE_DOCUMENT
import com.bogdan.kolomiiets.birthdayreminder.RequestCodes.Companion.INTENT_ACTION_GET_CONTENT
import com.bogdan.kolomiiets.birthdayreminder.RequestCodes.Companion.REQUEST_READ_EXTERNAL_STORAGE_FOR_IMPORT
import com.bogdan.kolomiiets.birthdayreminder.RequestCodes.Companion.REQUEST_WRITE_EXTERNAL_STORAGE_FOR_EXPORT
import com.bogdan.kolomiiets.birthdayreminder.adapters.EventsRecyclerViewAdapter
import com.bogdan.kolomiiets.birthdayreminder.models.Event
import com.bogdan.kolomiiets.birthdayreminder.models.Event.Companion.TYPE_HOLIDAY
import com.bogdan.kolomiiets.birthdayreminder.utils.*
import com.bogdan.kolomiiets.birthdayreminder.viewmodels.EventsViewModel
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.MobileAds
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.google.gson.reflect.TypeToken
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.io.InputStream
import java.io.OutputStream
import java.util.concurrent.Executors

class MainActivity : AppCompatActivity(), View.OnClickListener, OnPopUpMenuItemClick {
    private lateinit var newEventFab: FloatingActionButton
    private lateinit var noEvents: TextView
    private lateinit var eventViewModel: EventsViewModel
    private lateinit var eventRecyclerView: RecyclerView
    private lateinit var eventsRecyclerViewAdapter: EventsRecyclerViewAdapter
    private lateinit var searchView: SearchView
    private lateinit var disposable: Disposable
    private lateinit var adView: AdView
    private lateinit var jsonElement: JsonElement

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //init adView and adListener
        MobileAds.initialize(this) {}
        val adRequest = AdRequest.Builder().build()
        adView = findViewById(R.id.adView)
        adView.loadAd(adRequest)
        adView.adListener = object : AdListener() {

            override fun onAdClosed() {
                adView.loadAd(adRequest)
            }
        }

        //init UI components
        newEventFab = findViewById(R.id.new_event_fab)
        newEventFab.setOnClickListener(this)
        noEvents = findViewById(R.id.no_events)
        eventRecyclerView = findViewById(R.id.event_recycler)

        //init viewModel
        eventViewModel = ViewModelProviders.of(this).get(EventsViewModel::class.java)
        eventViewModel.events.observe(this, Observer {
            noEvents.isVisible = it.isNullOrEmpty()
            eventsRecyclerViewAdapter.setData(it)
        })

        //setup recyclerView
        eventRecyclerView.setHasFixedSize(true)
        eventRecyclerView.layoutManager = LinearLayoutManager(this)
        eventsRecyclerViewAdapter = EventsRecyclerViewAdapter(this)
        eventRecyclerView.adapter = eventsRecyclerViewAdapter

        //checking if it's first run
        val preferences = getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
        if (!preferences.contains(CHECKED_RADIO)) {
            preferences
                    .edit()
                    .putBoolean(IS_REMINDER_ON, true)
                    .putInt(CHECKED_RADIO, RADIO_SAME_DAY)
                    .putInt(REMINDER_HOUR, BASE_REMINDER_HOUR)
                    .putInt(REMINDER_MINUTE, BASE_REMINDER_MINUTE)
                    .apply()

            setEveryDayAlarm(this, BASE_REMINDER_HOUR, BASE_REMINDER_MINUTE, true)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_item, menu)
        searchView = (menu.findItem(R.id.item_search)).actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextChange(newText: String?): Boolean {
                eventViewModel.filter.postValue(newText ?: "")
                return true
            }

            override fun onQueryTextSubmit(query: String?): Boolean {
                return false
            }
        })
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.item_about -> showAboutDialog()
            R.id.item_export_to_file -> onMenuItemExportToFile()
            R.id.item_import_into_database -> onMenuItemImportIntoDatabase()
            R.id.item_settings -> startActivity(Intent(this, SettingsActivity::class.java))
        }
        return true
    }

    private fun onMenuItemExportToFile() =
            if (PermissionChecker.PERMISSION_GRANTED == PermissionChecker.checkSelfPermission(this, WRITE_EXTERNAL_STORAGE)) {
                if (isExternalStorageAvailable() and !isExternalStorageReadOnly()) {
                    disposable = eventViewModel.getEvents()
                            .subscribeOn(Schedulers.io())
                            .doOnSuccess { eventsList ->
                                if (eventsList.isNotEmpty()) {
                                    jsonElement = GsonBuilder().create().toJsonTree(eventsList)
                                    val filename = getString(R.string.file_name, getStringDate())
                                        intentCreateFile("*/*", filename)
                                } else {
                                    runOnUiThread { showToast(R.string.nothing_to_import) }
                                }
                            }
                            .doOnError { error: Throwable? -> showToast(error?.message ?: error.toString()) }
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe()
                } else showToast(R.string.external_storage_not_available)
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(WRITE_EXTERNAL_STORAGE), REQUEST_WRITE_EXTERNAL_STORAGE_FOR_EXPORT)
            }

    private fun intentCreateFile(mimeType: String, fileName: String) {
        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        intent.type = mimeType
        intent.putExtra(Intent.EXTRA_TITLE, fileName)
        startActivityForResult(intent, INTENT_ACTION_CREATE_DOCUMENT)
    }

    private fun onMenuItemImportIntoDatabase() {
        if (PermissionChecker.PERMISSION_GRANTED == PermissionChecker.checkSelfPermission(this, READ_EXTERNAL_STORAGE)) {
            val intent = Intent(Intent.ACTION_GET_CONTENT)
            intent.type = "*/*"
            intent.resolveActivity(packageManager)?.let {
                showToast(getString(R.string.select_a_file_for_import))
                startActivityForResult(intent, INTENT_ACTION_GET_CONTENT)
            }
        } else {
            ActivityCompat.requestPermissions(this, arrayOf(READ_EXTERNAL_STORAGE), REQUEST_READ_EXTERNAL_STORAGE_FOR_IMPORT)
        }
    }

    private fun executeExportDataToFile(resultIntent: Intent?) {
        resultIntent?.data?.let {
            contentResolver.openOutputStream(it).use { outputStream: OutputStream? -> outputStream?.write(jsonElement.toString().toByteArray())
            outputStream?.flush()
            }
        }
    }

        private fun executeImportDataIntoDatabase(resultIntent: Intent?) {
        resultIntent?.data?.let {
            Executors.newSingleThreadExecutor().execute {
                val stringBuilder = StringBuilder()
                contentResolver.openInputStream(it).use { inputStream: InputStream? -> inputStream?.bufferedReader()?.forEachLine { line -> stringBuilder.append(line) } }

                /**
                 * working code
                 * version 1
                 */
                /*val jsonArray = JSONArray(stringBuilder.toString())
                val events = mutableListOf<Event>()
                for (i in 0 until jsonArray.length()) {
                    events.add(Gson().fromJson(jsonArray[i].toString(), Event::class.java))
                }*/

                /**
                 * working code
                 * version 2
                 */
                val events = Gson().fromJson<List<Event>>(stringBuilder.toString(), object : TypeToken<List<Event>>() {}.type)

                if (!events.isNullOrEmpty()) {
                    //preparing for import
                    events.forEach { if (it.type == TYPE_HOLIDAY) it.phone = "" }
                    //execute import
                    eventViewModel.insertEvents(events)
                } else runOnUiThread { showToast(R.string.nothing_to_import) }
            }
        }
    }

    private fun showAboutDialog() {
        val dialogBuilder = AlertDialog.Builder(this)
        dialogBuilder
                .setIcon(R.mipmap.ic_launcher)
                .setMessage(R.string.about_message)
                .setTitle(R.string.app_name)
                .setPositiveButton(R.string.ok) { dialogInterface, _ -> dialogInterface.dismiss() }
        dialogBuilder.show()
    }

    override fun onClick(viev: View) {
        when (viev.id) {
            R.id.new_event_fab -> startActivity(Intent(this, NewOrUpdateEventActivity::class.java))
        }
    }

    override fun onBackPressed() {
        if (!searchView.isIconified) {
            searchView.onActionViewCollapsed()
        } else {
            super.onBackPressed()
        }
    }

    override fun deleteItemClick(eventId: Int) {
        eventViewModel.deleteEvent(eventId)
    }

    override fun changeItemClick(event: Event) {
        startActivity(Intent(this, NewOrUpdateEventActivity::class.java).putExtra(EVENT, event))
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (grantResults.isNotEmpty() and (grantResults[0] == PermissionChecker.PERMISSION_GRANTED)) {
            when (requestCode) {
                REQUEST_READ_EXTERNAL_STORAGE_FOR_IMPORT -> onMenuItemImportIntoDatabase()
                REQUEST_WRITE_EXTERNAL_STORAGE_FOR_EXPORT -> onMenuItemExportToFile()
            }
        } else super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == RESULT_OK) {
            when (requestCode) {
                INTENT_ACTION_GET_CONTENT -> executeImportDataIntoDatabase(data)
                INTENT_ACTION_CREATE_DOCUMENT -> executeExportDataToFile(data)
                else -> super.onActivityResult(requestCode, resultCode, data)
            }
        }
    }

    override fun onPause() {
        super.onPause()
        if (::disposable.isInitialized) {
            disposable.dispose()
        }
    }
}